lib/puppet/util.rb

```diff
@@ -349,6 +356,101 @@ def uri_to_path(uri)
   end
   module_function :uri_to_path
 
+  private
+
+  RFC_3986_URI_REGEX = /^(?<scheme>([^:\/?#]+):)?(?<authority>\/\/([^\/?#]*))?(?<path>[^?#]*)(\?(?<query>[^#]*))?(#(?<fragment>.*))?$/
+
+  public
+
+  # Percent-encodes a URI query parameter per RFC3986 - https://tools.ietf.org/html/rfc3986
+  #
+  # The output will correctly round-trip through URI.unescape
+  #
+  # @param [String query_string] A URI query string that may be in the form of:
+  #
+  #   query
+  #   query=foo
+  #   query=foo&query2=bar
+  #   query=foo+bar
+  #   query=hello world are you there
+  #
+  #   Note: Also usable by fragments, but not suitable for paths
+  #
+  # @return [String] a new string containing an encoded query string per the
+  #   rules of RFC3986.
+  #
+  #   In particular,
+  #   query will encode + as %2B and space as %20
+  def uri_query_encode(query_string)
+    return nil if query_string.nil?
+
+    # query can encode space to %20 OR +
+    # + MUST be encoded as %2B
+    # in RFC3968 both query and fragment are defined as:
+    # = *( pchar / "/" / "?" )
+    # CGI.escape turns space into + which is the most backward compatible
+    # however it doesn't roundtrip through URI.unescape which prefers %20
+    CGI.escape(query_string).gsub('+', '%20').gsub('%3D', '=').gsub('%26', '&')
```

Comment:
The decoding of `%26` to `&` has caused The problem reported in https://tickets.puppetlabs.com/browse/PUP-7658.

`uri_query_encode` conceptually works over an entire query string - and unencodes any `&` that may have been encoded, though this may be a bit naive.  

Within `fact_handler.rb`, all of the fact data should be encoded, else it injects delimiters into values that end up being URI percent encoded OR form-encoded.  The correct helper to use there is actually [URI.encode_www_form_component(str)](https://ruby-doc.org/stdlib-1.9.3/libdoc/uri/rdoc/URI.html#method-c-encode_www_form_component)